src/main.rs

```diff
@@ -40,6 +41,62 @@ fn find_message_pda(wormhole_pid: &Pubkey, slot: u64) -> Pubkey {
     .0
 }
 
+const FAILED_TO_DECODE: &str = "Failed to decode account data";
+const INVALID_UNRELIABLE_DATA_FORMAT: &str = "Invalid unreliable data format";
+
+#[derive(Debug)]
+enum VerifyUpdateError {
+    InvalidMessagePDA,
+    InvalidEmitterChain,
+    InvalidAccumulatorAddress,
+    #[allow(dead_code)]
+    DecodingError(String),
+}
+
+fn decode_and_verify_update(
+    wormhole_pid: &Pubkey,
+    accumulator_address: &Pubkey,
+    update: Response<RpcKeyedAccount>,
+) -> Result<PostedMessageUnreliableData, VerifyUpdateError> {
+    if find_message_pda(wormhole_pid, update.context.slot).to_string() != update.value.pubkey {
+        return Err(VerifyUpdateError::InvalidMessagePDA);
+    }
+    let data = update
+        .value
+        .account
+        .data
+        .decode()
+        .ok_or(VerifyUpdateError::DecodingError(
+            FAILED_TO_DECODE.to_string(),
+        ))?;
+    let unreliable_data: PostedMessageUnreliableData =
+        BorshDeserialize::deserialize(&mut data.as_slice()).map_err(|e| {
+            VerifyUpdateError::DecodingError(format!("{}: {}", INVALID_UNRELIABLE_DATA_FORMAT, e))
+        })?;
+
+    if Chain::Pythnet != unreliable_data.emitter_chain.into() {
+        return Err(VerifyUpdateError::InvalidEmitterChain);
+    }
+
+    if accumulator_address != &Pubkey::from(unreliable_data.emitter_address) {
+        return Err(VerifyUpdateError::InvalidAccumulatorAddress);
+    }
+
+    Ok(unreliable_data)
+}
+
+fn new_body(unreliable_data: &PostedMessageUnreliableData) -> Body<&RawMessage> {
```

Comment:
hmmm the name is a bit confusing here. i think `message_data_to_body` is a better name here. (pro tip: probably best is implementing From trait for Body from UnreliableData but then you'd need to use Body<Box<RawMessage>>)

p.s: posted_message.rs seems like a copy paste and i always forget it. can you add a module comment on top of that explain what that is and link to the reference implementation?